webapp/static/scripts/remote/controller.js

```diff
@@ -0,0 +1,220 @@
+function mouseStartOnSliders(e, obj) {
+    obj.stick.manip = true;
+    getMousePosOnSliders(e, obj);
+    e.preventDefault();// prevent canceling this event
+}
+
+function mouseMoveOnSliders(e, obj) {
+    if(e.buttons == 1){
+        getMousePosOnSliders(e, obj);
+    }    
+    e.preventDefault();
+}
+
+function mouseEndOnSliders(e, obj) {
+    obj.stick.manip = false;
+    obj.value = 0;
+}
+
+function getMousePosOnSliders(e, obj) {
+    let targetRect = e.target.getBoundingClientRect();
+    const mouseX = ((e.pageX - targetRect.left) / obj.width * 2 - 1) / ((obj.width - obj.stick.radius * 2) / obj.width) * 100;
+    const mouseY = ((e.pageY - targetRect.top) / obj.height * -2 + 1) / ((obj.height - obj.stick.radius * 2) / obj.height) * 100;
+    obj.value = obj.horizontal ? mouseX : mouseY;
+    // console.log("slider value:", obj.value);
+}
+
+
+// get data from physical gamepads
+// Google's Chrome handles gamepads differently than others, so we implement a workaraound
+function getGamepadChanges() {
+    // do this here as it is needed for Chrome workaround
+    gamepads = navigator.getGamepads();
+    /*  according to the "standard" mapping scheme
+     *  (compatible w/ xBox 360 & other xinput controllers):
+     * axis[0] = left stick X axis
+     * axis[1] = left stick Y axis
+     * axis[2] = right stick X axis
+     * axis[3] = right stick Y axis
+     */
+    let result = [];
+    // only grab data from gamepad @ index 0
+    if (gamepads.length) {
+        if (gamepads[0] != null) {// Chrome specific workaround
+            // if there is at least 3 axes on the gamepad
+            if (gamepads[0].axes.length >= 3){
+                // using a deadzone of +/- 4%
+                if (gamepads[0].axes[1] > 0.04 || gamepads[0].axes[1] < -0.04){
+                    result.push(gamepads[0].axes[1] * -1); // used for speed
+                }
+                else{ // axis is within deadzone
+                    result.push(0);
+                }
+                if (gamepads[0].axes[2] > 0.04 || gamepads[0].axes[2] < -0.04) {
+                    result.push(gamepads[0].axes[2]); // used for turn
+                }
+                else{ // axis is within deadzone
+                    result.push(0);
+                }
+            }
+        /*
+        for (i = 0; i < gamepads.length; i++){
+            // show axes data
+            for (j = 0; j < gamepads[i].axes.length; j++) {
+                var temp = gamepads[i].axes[j];
+                if (temp > 0.025 || temp < -0.025) {
+                    console.log("gamepad[" + i + "], axis[" + j + "] = " + temp);
+                }
+            }
+            //show button presses (analog triggers have preset threshold)
+            for (j = 0; j < gamepads[i].buttons.length; j++) {
+                var temp = gamepads[i].buttons[j].pressed;
+                if (temp)
+                console.log("gamepad[" + i + "], button[" + j + "] = " + temp);
+            }
+        }
```

Comment:
Do you plan to remove this testing code? If not, I'd recommend uncommenting it and putting it behind a boolean flag.

Reply:
I'm saving it there (online resources relating to gamepads used in browsers are rather sparce) so we know how to access the gamepad buttons when we need to implement something like brakes (or even triggering an LED sequence). I guess I could remove the section that displays all the axes...